cmsl1t/collections/vectorized.py

```diff
@@ -0,0 +1,167 @@
+import awkward
+from collections import defaultdict
+import logging
+import numpy as np
+import random
+from rootpy.plotting import Hist
+
+from . import BaseHistCollection
+from ..utils.iterators import pairwise
+from .. import PY3
+from ..io import to_root
+
+logger = logging.getLogger(__name__)
+
+
+def extend(arr1, starts, stops):
+    repeat = stops - starts
+    return np.repeat(arr1, repeat, axis=0)
+
+
+def split_input(inner_indices, x, w):
+    content = x
+    if hasattr(x, 'starts'):
+        inner_indices = extend(inner_indices, x.starts, x.stops)
+        content = x.content
+
+    for u in np.unique(inner_indices):
+        mask = inner_indices == u
+        if not isinstance(mask, (list, np.ndarray)):
+            mask = np.array([mask])
+        yield u, content[mask], w[mask]
+
+
+class VectorizedHistCollection(BaseHistCollection):
+
+    def __init__(self, innerBins, innerLabel='inner', **kwargs):
+        # if we want to generalize to N dim, innerBins needs to be an array of innerBins
+        # TODO: last dimension should probably be a normal dictionary
+        dimensions = kwargs.pop('dimensions', 2)
+        self._name = kwargs.pop('name', str(hex(random.getrandbits(128)))[2:10])
+        self._execute_before_write = kwargs.pop('execute_before_write', [])
+        if PY3:
+            super(VectorizedHistCollection, self).__init__(dimensions)
+        else:
+            BaseHistCollection.__init__(self, dimensions)
+
+        self._innerBins = innerBins
+        self._innerLabel = innerLabel
+        self._innerHist = Hist(100, 0, 100, name=innerLabel + '_' + self._name)
```

Comment:
Where do the `100`s come from?  They seem like magic! :smile:

Reply:
Argh, yes, a speck of the magical past that is debugging